src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataLocation.cs

```diff
@@ -31,102 +31,51 @@ internal sealed class DiagnosticDataLocation
         public readonly TextSpan? SourceSpan;
 
-        [DataMember(Order = 3)]
-        public readonly int OriginalStartLine;
-
-        [DataMember(Order = 4)]
-        public readonly int OriginalStartColumn;
-
-        [DataMember(Order = 5)]
-        public readonly int OriginalEndLine;
-
-        [DataMember(Order = 6)]
-        public readonly int OriginalEndColumn;
-
         /// <summary>
-        /// Null if path is not mapped and <see cref="OriginalFilePath"/> contains the actual path.
-        /// Note that the value might be a relative path. In that case <see cref="OriginalFilePath"/> should be used
-        /// as a base path for path resolution.
+        /// Null if path is not mapped and <see cref="OriginalFileSpan"/> contains the actual path. Note that the value
+        /// might be a relative path. In that case <see cref="OriginalFileSpan"/> should be used as a base path for path
+        /// resolution.
         /// </summary>
-        [DataMember(Order = 7)]
-        public readonly string? MappedFilePath;
-
-        [DataMember(Order = 8)]
-        public readonly int MappedStartLine;
-
-        [DataMember(Order = 9)]
-        public readonly int MappedStartColumn;
-
-        [DataMember(Order = 10)]
-        public readonly int MappedEndLine;
-
-        [DataMember(Order = 11)]
-        public readonly int MappedEndColumn;
+        [DataMember(Order = 3)]
+        public readonly FileLinePositionSpan? MappedFileSpan;
 
         public DiagnosticDataLocation(
-            string originalFilePath,
+            FileLinePositionSpan originalFileSpan,
             DocumentId? documentId = null,
             TextSpan? sourceSpan = null,
-            int originalStartLine = 0,
-            int originalStartColumn = 0,
-            int originalEndLine = 0,
-            int originalEndColumn = 0,
-            string? mappedFilePath = null,
-            int mappedStartLine = 0,
-            int mappedStartColumn = 0,
-            int mappedEndLine = 0,
-            int mappedEndColumn = 0)
+            FileLinePositionSpan? mappedFileSpan = null)
         {
-            Contract.ThrowIfNull(originalFilePath);
+            Contract.ThrowIfNull(originalFileSpan.Path);
 
-            OriginalFilePath = originalFilePath;
+            OriginalFileSpan = originalFileSpan;
             DocumentId = documentId;
             SourceSpan = sourceSpan;
-            MappedFilePath = mappedFilePath;
-            MappedStartLine = mappedStartLine;
-            MappedStartColumn = mappedStartColumn;
-            MappedEndLine = mappedEndLine;
-            MappedEndColumn = mappedEndColumn;
-            OriginalStartLine = originalStartLine;
-            OriginalStartColumn = originalStartColumn;
-            OriginalEndLine = originalEndLine;
-            OriginalEndColumn = originalEndColumn;
+            MappedFileSpan = mappedFileSpan;
+
+            if (MappedFileSpan is { HasMappedPath: false })
+                MappedFileSpan = null;
```

Comment:
this ensures a nice invariant that MappedFileSpan is only non-null when there is actual mapping going on.